### PR TITLE
Implement Glowing API

### PR DESCRIPTION
**Pull request by ZephireNZ**
*Continuation of #696*
API: SpongePowered/SpongeAPI#1279

This implements the new 1.9 Glowing API

Tests:

```
@Plugin(id="test-plugin", name="TestPlugin", version="1.0")
public class TestPlugin {

    @Inject
    private Logger log;

    @Listener
    public void onSpawn(SpawnEntityEvent event) {
        for(Entity entity : event.getEntities()) {
            if(entity instanceof Projectile) return;
            entity.offer(Keys.GLOWING, true);
            log.debug(entity.getUniqueId().toString() + " glowing: " + entity.get(Keys.GLOWING));
        }
    }

    @Listener
    public void onBow(SpawnEntityEvent event) {
        for(Entity entity : event.getEntities()) {
            if(!(entity instanceof Projectile)) return;
            GlowingData data = entity.getOrCreate(GlowingData.class).get();
            log.debug("projectile glowing before: " + data.glowing().get());
            data.set(Keys.GLOWING, true);
            entity.offer(data);
            Sponge.getScheduler().createTaskBuilder()
                    .delayTicks(0)
                    .execute(() -> {
                        GlowingData glowingData = entity.get(GlowingData.class).get();
                        log.debug("projectile glowing after: " + glowingData.glowing().get());
                    }).submit(this);
        }
    }

}
```

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeGlowingData.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package org.spongepowered.common.data.manipulator.immutable.entity;
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeGlowingData.java]: Need license header

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeGlowingData.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import org.spongepowered.common.data.manipulator.mutable.common.AbstractBooleanData;
+import org.spongepowered.common.data.util.DataConstants;
+
+public class SpongeGlowingData extends AbstractBooleanData<GlowingData, ImmutableGlowingData> implements GlowingData {
+
+    public SpongeGlowingData(boolean value) {
+        super(GlowingData.class, value, Keys.GLOWING, ImmutableSpongeGlowingData.class, DataConstants.DEFAULT_GLOWING);
+    }
+
+    public SpongeGlowingData() {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeGlowingData.java]: Minor, but move this above the other constructor (least args to most args).

File: src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    protected boolean set(Entity dataHolder, Boolean value) {
+        dataHolder.setGlowing(checkNotNull(value, "value"));
+        return true;
+    }
+
+    @Override
+    protected Optional<Boolean> getVal(Entity dataHolder) {
+        return OptBool.of(dataHolder.isGlowing());
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java]: You can use `OptBool` here.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    protected ImmutableValue<Boolean> constructImmutableValue(Boolean value) {
+        return ImmutableSpongeValue.cachedOf(Keys.GLOWING, value, DataConstants.DEFAULT_GLOWING);
+    }
+
+    @Override
+    protected Value<Boolean> constructValue(Boolean actualValue) {
+        return new SpongeValue<>(Keys.GLOWING, DataConstants.DEFAULT_GLOWING, actualValue);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java]: `this.key` (or `this.usedKey`, one of the two) can be used here.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    protected Optional<Boolean> getVal(Entity dataHolder) {
+        return OptBool.of(dataHolder.isGlowing());
+    }
+
+    @Override
+    protected ImmutableValue<Boolean> constructImmutableValue(Boolean value) {
+        return ImmutableSpongeValue.cachedOf(Keys.GLOWING, value, DataConstants.DEFAULT_GLOWING);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java]: `this.key` (or `this.usedKey`, one of the two) can be used here.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java (hunk 3 above)
Comment (ZephireNZ) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java]: ... but this is using OptBool?

Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/GlowingDataProcessor.java]: My eyes appear to have played a trick on me when I commented on that. :smile_cat:

**Comment by ZephireNZ** [comment]
@gabizou, this is all ready to go as far as I can tell. Even got a test in there ^.^

**Pull request merged**